### PR TITLE
feat: Add support for case conditional function in OData URI parser

### PR DESCRIPTION
**Pull request by xuzhg**
This commit introduces comprehensive support for the 'case' conditional function in OData URI queries, enabling conditional expressions in $compute, $filter, and $orderby query options.

### Core Implementation
- Added case function support in FunctionCallBinder with type inference logic
- Implemented CreateCaseFunctionNode method for binding case expressions
- Added type promotion logic for homogeneous result types
- Support for mixed types (returns Untyped)
- Proper handling of String, Boolean, Collection, and Untyped return types

### Syntax Support
- Basic case syntax: case(condition: result, ..., true: default)
- Multiple condition/result pairs (tested up to 6 pairs)
- Compound conditions with AND operator
- Collection operations ()
- Null literal support and null checking
- Empty collection literals ([])

### Query Option Integration
- $compute: Computed properties using case expressions
- $filter: Case expressions as filter predicates
- $orderby: Case expressions for custom sort order (ASC/DESC)
- $select: Selection of case-computed properties

### CaseConditionalFunctionTests.cs (13 new tests)
- Basic Functionality: Collection fallback, email selection, age categorization, mixed properties
- Edge Cases: Single pair, default-only, many pairs (6+)
- Integration: Filter expressions, boolean filters, orderby (ASC/DESC)
- Null Handling: Null results, null checks with ConvertNode support

### SelectExpandFunctionalTests.cs (1 updated test)
- Case in  with  integration

## Examples

`odata
// Age categorization
?=case(Age lt 18: 'Minor', true: 'Adult') as AgeGroup

// Priority-based filtering
?=case(Status eq 'Active': true, true: false)

// Custom sort order
?=case(Status eq 'VIP': 0, Status eq 'Premium': 1, true: 2)

// Collection fallback
?=case(CustomTags/ gt 0: CustomTags, true: SystemTags) as SearchTags

// Null-safe display
?=case(Name ne null: Name, true: 'Unknown') as SafeName `

## Type Inference Rules
- All string results → String type
- All boolean results → Boolean type
- All numeric results → Numeric type with promotion
- Mixed types or collections → Untyped

## Breaking Changes
None. This is a new feature addition.

## Related Issues
Implements case conditional function per OData specification.

### Issues

*This pull request fixes #xxx.*

### Description

*Briefly describe the changes of this pull request.*

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### Repository notes

Team members can start a CI build by adding a comment with the text `/AzurePipelines run` to a PR. A bot may respond indicating that there is no pipeline associated with the pull request. This can be ignored if the build is triggered.

Team members should **not** trigger a build this way for pull requests coming from forked repositories. They should instead trigger the build manually by setting the "branch" to `refs/pull/{prId}/merge` where `{prId}` is the ID of the PR.

### CODE REVIEW

**Comment by xuzhg** [comment]
@copilot Please link the related issue to this PR.

File: src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs [hunk 1]
```diff
@@ -855,5 +861,98 @@ private static IEdmTypeReference TryGetTypeReference(IEdmModel model, string ful
...
+            // Verify the condition/result:
+            IEdmTypeReference returnType = null;
+            for (int i = 0; i < args.Count; i += 2)
+            {
+                SingleValueNode singleValueNode = args[i] as SingleValueNode;
+
+                // the first expression – the condition – MUST be a Boolean expression.
+                if (singleValueNode == null || !singleValueNode.TypeReference.IsBoolean())
+                {
+                    throw new ODataException(Error.Format(SRResources.FunctionCallBinder_CaseConditionMustBeSingleBooleanValue, i % 2));
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs]: #resolved

File: src/Microsoft.OData.Core/UriParser/TypePromotionUtils.cs [hunk 2]
```diff
@@ -551,6 +551,11 @@ internal static bool CanConvertTo(SingleValueNode sourceNodeOrNull, IEdmTypeRefe
             return MetadataUtilsCommon.CanConvertPrimitiveTypeTo(sourceNodeOrNull, sourcePrimitiveTypeReference.PrimitiveDefinition(), targetPrimitiveTypeReference.PrimitiveDefinition());
         }
 
+        internal static bool IsNumericType(IEdmTypeReference type)
+        {
+            return GetNumericTypeKind(type) != NumericTypeKind.NotNumeric;
+        }
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/UriParser/TypePromotionUtils.cs:557]: Won't fix

File: test/UnitTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/CaseConditionalFunctionTests.cs [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+            var exception = Assert.Throws<ODataException>(() =>
+                ParseCompute(compute, this.model, this.userType, this.usersSet));
+
+            // The error should mention even number requirement or parameter count
+            Assert.Equal("':' expected at position 49 in 'case(Age lt 18: 'Minor', Age lt 65: 'Adult', true) as Invalid'. The 'case' function requires 'condition:result' pairs.", exception.Message);
+        }
+
+        /// <summary>
+        /// Tests that case expression with non-boolean condition throws appropriate error.
+        /// Query uses a string literal where a boolean condition is expected.
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply test/UnitTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/CaseConditionalFunctionTests.cs:561]: Won't fix

File: src/Microsoft.OData.Core/UriParser/SemanticAst/SingleValueFunctionCallNode.cs [hunk 4]
```diff
@@ -79,7 +79,7 @@ public SingleValueFunctionCallNode(string name, IEnumerable<IEdmFunction> functi
             if (returnedTypeReference != null)
             {
                 if (returnedTypeReference.IsCollection()
-                    || !(returnedTypeReference.IsComplex() || returnedTypeReference.IsPrimitive() || returnedTypeReference.IsEnum()))
+                    || !(returnedTypeReference.IsComplex() || returnedTypeReference.IsPrimitive() || returnedTypeReference.IsEnum() || returnedTypeReference.IsUntyped()))
                 {
                     throw new ArgumentException(SRResources.Nodes_SingleValueFunctionCallNode_ItemTypeMustBePrimitiveOrComplexOrEnum);
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/UriParser/SemanticAst/SingleValueFunctionCallNode.cs:84]: Won't fix
